Business logic updates for IPP feedback survey banner

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->
#### Business logic updates for IPP feedback survey banner

**Note:** This PR is part of the IPP in-app feedback banner project https://github.com/woocommerce/woocommerce-android/issues/8116. It targets the feature branch feature/8116-ipp-feedback-banner.

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR introduces the following new use case classes:
* `MarkFeedbackBannerAsDismissed`
* `MarkFeedbackBannerAsDismissedForever`
* `MarkIPPFeedbackSurveyAsCompleted`

that are responsible for storing banner actions in `SharedPreferences`.

It also adds the following updates to existing use cases:
* `GetIPPFeedbackBannerData` - renamed from `GetIPPFeedbackSurveyUrl` and modified to return banner message along with survey url. These values are wrapped up by `IPPFeedbackBanner ` data class.
* `ShouldShowFeedbackBanner` - updated to limit showing the banner only to stores in CA and US.

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

There is no UI part associated with the use cases yet, so I just recommend reviewing unit tests: `ShouldShowFeedbackBannerTest` and `GetIPPFeedbackBannerDataTest` and make sure they are complete.

### Roadmap
In the next PR, I'll add a feature flag and UI for the banner.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->


- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Show feedback banner only to stores in CA&US

samiuelson committed:
Add use cases for storing banner actions in prefs

samiuelson committed:
Update `GetIPPFeedbackSurveyUrl` to return banner messages

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/feedback/ipp/GetIPPFeedbackBannerData.kt:
```diff
@@ -11,15 +13,18 @@ import org.wordpress.android.fluxc.store.WCInPersonPaymentsStore
 
-class GetIPPFeedbackSurveyUrl @Inject constructor(
+/**
+ * Use case class to detect specific user group type and return the IPP feedback banner data: message and survey url.
```
np: I would avoid any comments that describe the code. Unless they explain why not what

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/feedback/ipp/GetIPPFeedbackBannerData.kt:
```diff
@@ -11,15 +13,18 @@ import org.wordpress.android.fluxc.store.WCInPersonPaymentsStore
     @Suppress("ReturnCount")
     @Throws(IllegalStateException::class)
-    suspend operator fun invoke(): String {
+    suspend operator fun invoke(): IPPFeedbackBanner {
```
np: Not this line and not this PR, but might be useful

```
throw IllegalStateException("Failed to fetch transactions summary")
```

maybe become

```
checkNotNull(response.result?.transactionsCount) { "Transactions count must not be null" }
```

And then idea won't complain on these lines

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/feedback/ipp/GetIPPFeedbackBannerData.kt:
```diff
@@ -80,6 +90,15 @@ class GetIPPFeedbackSurveyUrl @Inject constructor(
         @VisibleForTesting
         const val SURVEY_URL_IPP_NINJA = "https://woocommerce.com/ninja"
 
+        @VisibleForTesting
```
Wouldn't it be better to assert the actual values in the tests? Currently, you test that the result matches the constant in the code, so if let's say there is a mistaken change that brings a typo in these values, the tests won't catch that. Also, you won't need these `VisibleForTesting` annotations

samiuelson replied:
I agree, I made constants private. I left tests related to checking string res IDs, because it's not possible to access resources in the unit tests.

kidinov replied:
> because it's not possible to access resources in the unit tests.

Why? `R.java` is just a generated file. We have a lot of tests that use resource ids or I got you wrong?

samiuelson replied:
For some reason, my IDE couldn't resolve `android.R` class in the unit tests module. This turned out to be working good from commandline. Thanks for the hint.

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/feedback/ipp/ShouldShowFeedbackBanner.kt:
```diff
@@ -15,19 +17,28 @@ class ShouldShowFeedbackBanner @Inject constructor(
-     * 2. Check if WCPay plugin is used if no, return false.
-     * 3. Check if survey has been already completed, if yes return false.
-     * 4. Check if survey has been dismissed forever, if yes return false,
+     * 1. Check if store's country code is US or CA, if not, return false
```
Is this comment any better than the code inside the function?

samiuelson replied:
Agreed, this was useful only during development. Removed.

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/feedback/ipp/ShouldShowFeedbackBanner.kt:
```diff
@@ -46,5 +57,6 @@ class ShouldShowFeedbackBanner @Inject constructor(
 
     private companion object {
         private const val REMIND_LATER_INTERVAL_DAYS = 7
+        val SUPPORTED_COUNTRIES = listOf("US", "CA")
```
it might be beneficial to use `CardReaderConfigForSupportedCountry` to determine what countries we support for IPP, so we won't be hunting all over the code if another country will be added

kidinov replied:
You can use `CardReaderConfigFactory` for that

samiuelson replied:
Thanks for the suggestion. We're using `CardReaderConfigFactory` now.

kidinov commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/payments/feedback/ipp/GetIPPFeedbackBannerDataTest.kt:
```diff
@@ -132,7 +135,61 @@ class GetIPPFeedbackSurveyUrlTest : BaseUnitTest() {
+    @Test
+    fun `given banner should be displayed and user is a newbie, then should display correct message`() = runBlocking {
+        // given
+        whenever(shouldShowFeedbackBanner()) doReturn true
```
I think for more consistency we might want to use `.thenReturn(true)`. There are usages of `doReturn` in the code base but it's just a few files

samiuelson replied:
Okay, switched to `thenReturn`.

kidinov reviewed (commented):
Thank you for working on that! I've just left a few minor comments. Please take a look

samiuelson committed:
Improve `ShouldShowFeedbackBanner` code

samiuelson committed:
Improve `GetIPPFeedbackBannerDataTest` and `ShouldShowFeedbackBannerTest` code style

samiuelson committed:
Clean up code

samiuelson committed:
Remove comments

samiuelson commented:
Thanks for the review @kidinov! I made improvements and I will appreciate another round of review.

kidinov reviewed (approved):
LGTM! Before merging I would just remove the rest of the `@VisibleForTesting` annotations and test resources' ids in the test directly, unless I am missing something

samiuelson committed:
Update unit `GetIPPFeedbackBannerDataTest`

Check against android.R.string resource ids instead of their handles form `GetIPPFeedbackBannerData` class.
